### DISCUSSION TITLE
Change page content from middleware.js

### DISCUSSION THREAD

**Discussion by jofri** [vercel/next.js#49086] [Help]
### Summary

Hi there - I've seen this great function from Netlify, allowing me to change the content on a rendered page through Next.js middleware.

[https://docs.netlify.com/integrations/frameworks/next-js/middleware/#middlewareresponse-object](https://docs.netlify.com/integrations/frameworks/next-js/middleware/#middlewareresponse-object)

Is there a way of achieving this without using Netlify's package?

Thanks!

**Comment by icyJoseph** [comment]
Do you mean like this? https://nextjs.org/docs/advanced-features/middleware#producing-a-response

```ts
// middleware.ts
import { NextRequest, NextResponse } from 'next/server'
import { isAuthenticated } from '@lib/auth'

// Limit the middleware to paths starting with `/api/`
export const config = {
  matcher: '/api/:function*',
}

export function middleware(request: NextRequest) {
  // Call our authentication function to check the request
  if (!isAuthenticated(request)) {
    // Respond with JSON indicating an error message
    return new NextResponse(
      JSON.stringify({ success: false, message: 'authentication failed' }),
      { status: 401, headers: { 'content-type': 'application/json' } }
    )
  }
}
```

Available since 13.1

**Reply by jofri** [reply]
Unfortunately, this does not let me access the content (body) that is sent to the client. Let's say it sends a page with a H1 tag that says "Hello world!". I'd like to intercept this code and change the text. This can be done with the Netlify function I linked previously.

**Reply by icyJoseph** [reply]
Yeah, that's probably not 100% doable. I'd think of rewriting to a Next page from middleware, something like `/_my_hidden/path`, and also forbid direct access to it.

That way, at least in the pages dir, you could control the result, but it is not what you want, like, this hidden path thing would need to account for absolutely every page combo you have... that's a no go most likely.

I reckon this feature is using something similar to the the lambda at edge model, where there's a viewer response step, that allows you to modify the response as it goes back to the viewer, perhaps it uses the origin response, either way, it is part of the "return" branch, whereas the regular Next.js middleware is part of the "incoming" branch.

Neat feature I must say, perhaps something Vercel should look into.

**Comment by jofri** [comment]
Thanks - yea that makes sense. Might suggest as a new Next feature!

**Comment by nickgrealy** [comment]
I was able to workaround this, by performing an explicit fetch for my backend content.

Example here -> https://github.com/orgs/vercel/discussions/3874#discussioncomment-6899400

**Reply by Ragnarrlothbrok** [reply]
I am also trying to achieve the same thing, but it is affecting the response time heavily, Is the same page performant for you while using this middleware approach?

**Reply by nickgrealy** [reply]
It basically doubles the load time, plus you don't get the benefits of streaming.

In my case it's a last resort, and saves me having to run a backend server. 

Cache Control (max-age + stale-while-revalidate) and small payloads have helped me reduce the effects of this.

**Reply by vikaskrsharma** [reply]
@nickgrealy above link is giving 404. can you please share the updated link if you have. i have a similar usecase and needed to modify the head tag on run time
